crates/engine/tree/src/tree/payload_processor/multiproof.rs

```diff
@@ -570,25 +592,53 @@ where
 
     /// Handles state updates.
     ///
-    /// Returns proof targets derived from the state update.
-    fn on_state_update(
-        &mut self,
-        source: StateChangeSource,
-        update: EvmState,
-        proof_sequence_number: u64,
-    ) {
+    /// Returns a number of proofs that were spawned.
+    fn on_state_update(&mut self, source: StateChangeSource, update: EvmState) -> u64 {
         let hashed_state_update = evm_state_to_hashed_post_state(update);
-        let proof_targets = get_proof_targets(&hashed_state_update, &self.fetched_proof_targets);
-        self.fetched_proof_targets.extend_ref(&proof_targets);
+        // Split the state update into already fetched and not fetched according to the proof
+        // targets.
+        let (fetched_state_update, not_fetched_state_update) =
+            hashed_state_update.partition_by_targets(&self.fetched_proof_targets);
+
+        let mut state_updates = 0;
+        // Send the state update with already fetched proof targets as one update.
+        if !fetched_state_update.is_empty() {
+            let _ = self.tx.send(MultiProofMessage::EmptyProof {
```

Comment:
this isn't obvious why do we check here for ! not empty

shouldn't this check for not_fetched (missing)